napari/animation/animate.py

```diff
@@ -0,0 +1,344 @@
+            states_dict.append(x)
+            # do not add frames after last key-frame
+            if ind < len(self.key_frames) - 1:
+                # do not add frames when switching camera
+                if (
+                    self.key_frames[ind]['ndisplay']
+                    == self.key_frames[ind + 1]['ndisplay']
+                ):
+                    for y in range(inter_steps[ind]):
+                        states_dict.append(copy.deepcopy(empty))
+        for ind, x in enumerate(states_dict):
+            x["frame"] = ind
+        self.states_dict = states_dict
+
+    def create_steps(self):
+        """Interpolate states between key-frames"""
+
+        self.create_state_dict()
+        self.interpolated_states = util.interpolate(self.states_dict)
+
+    def key_interpolframe(self, viewer):
+        """Progress through interpolated frames"""
+
+        self.create_steps()
+
+        new_frame = (self.current_interpolframe + 1) % len(self.states_dict)
+        self.update_viewer_from_state(new_frame)
+        self.current_interpolframe = new_frame
+
+    def update_viewer_from_state(self, frame):
+        """Set the viewer to a given interpolated frame
+
+        Parameters
+        -------
+        frame : int
+            frame to visualize
+        """
+
+        # set view type 2D/3D and camera state
+        self.viewer.dims.ndisplay = self.interpolated_states["ndisplay"][frame]
+        self.viewer.window.qt_viewer.view.camera.set_state(
+            self.interpolated_states["camera"][frame]
+        )
+
+        # assign interpolated visibility state
+        for j in range(len(self.viewer.layers)):
+            self.viewer.layers[j].visible = self.interpolated_states["vis"][
+                frame
+            ][j]
+
+        # adjust slider positions
+        for i in range(self.interpolated_states["sliders"].shape[1]):
+            self.viewer.dims.set_point(
+                i, self.interpolated_states["sliders"][frame][i]
+            )
+
+        # update view
+        self.viewer.window.qt_viewer.view.camera.view_changed()
+
+    def frame_generator(self, frame=None, with_viewer=False, max_shape=None):
+        """Generator of frames of the animation
+
+        Parameters
+        -------
+        frame : int
+            Specific frame to return.
+        with_viewer : bool
+            If True includes the napari viewer, otherwise just includes the
+            canvas.
+        max_shape : int
+            Size to which to rescale the largest axis. Only used without
+            viewer (with_viewer = False).
+
+        """
+
+        # create states
+        self.create_steps()
+
+        # capture SceneCanvas size of frist frame to set size of next ones
+        self.update_viewer_from_state(0)
+        frame_size = self.viewer.window.qt_viewer.canvas.size
+        # if no max size provided, take current size as default
+        if max_shape is None:
+            max_shape = max(frame_size)
```

Comment:
Following my comment below on using [skimage resize](https://scikit-image.org/docs/stable/api/skimage.transform.html#skimage.transform.resize) in the `screenshot` method I think we can drop the `frame_size = self.viewer.window.qt_viewer.canvas.size` and also drop the `if max_shape is None` from inside the generator.

I think we will want to add an extra line into `animate` though